### ISSUE TITLE
corrupted git repo?

### ISSUE THREAD

**Issue by SEJeff** [gorakhargosh/watchdog#871]
When trying to clone this repo as a dependency of YouCompleteMe, I'm getting this:
```
$ git clone https://github.com/gorakhargosh/watchdog
Cloning into 'watchdog'...
remote: Enumerating objects: 7582, done.
remote: Counting objects: 100% (410/410), done.
remote: Compressing objects: 100% (252/252), done.
error: object 0bc9713e4ae1664648f1b3fd1915638f28d66236: badTimezone: invalid author/committer line - bad time zone
fatal: fsck error in packed object
fatal: fetch-pack: invalid index-pack output
```

I've tried with these versions of git on different clients:
```
$ git --version
git version 2.30.2
```
and
```
$ git --version
git version 2.31.0
```

Looking at [that commit](https://github.com/gorakhargosh/watchdog/commit/0bc9713e4ae1664648f1b3fd1915638f28d66236.patch), it does show a very strange UTC offset of +51800 vs say a more [normal commit](https://github.com/gorakhargosh/watchdog/commit/77e1f4668bc42b4a8706b181d882fd68c05ab8b9.patch) with a UTC offset of -0600.

Is there some obvious magic to make this work?

**Comment by SEJeff** [comment]
Mitigated via updating my git config to `transfer.fsckobjects=false`, but still seems like a bad thing.

**Comment by taleinat** [comment]
I just cloned this repo today and it worked as expected.

Ubuntu 20.04, git version 2.25.1

**Comment by taleinat** [comment]
> Looking at [that commit](https://github.com/gorakhargosh/watchdog/commit/0bc9713e4ae1664648f1b3fd1915638f28d66236.patch), it does show a very strange UTC offset of +51800 vs say a more [normal commit](https://github.com/gorakhargosh/watchdog/commit/77e1f4668bc42b4a8706b181d882fd68c05ab8b9.patch) with a UTC offset of -0600.

Other commits in this repo have the same UTC offset, so I don't think that's the issue.

**Comment by SEJeff** [comment]
> > Looking at [that commit](https://github.com/gorakhargosh/watchdog/commit/0bc9713e4ae1664648f1b3fd1915638f28d66236.patch), it does show a very strange UTC offset of +51800 vs say a more [normal commit](https://github.com/gorakhargosh/watchdog/commit/77e1f4668bc42b4a8706b181d882fd68c05ab8b9.patch) with a UTC offset of -0600.
> 
> Other commits in this repo have the same UTC offset, so I don't think that's the issue.

I think it is the issue, but feel free to close this. The timezone is totally bogus, as prov. The workaround is to not enable `transfer.fsckobjects` when cloning this repo because it has invalid metadata.

```
$ git config --global transfer.fsckobjects true

$ git clone https://github.com/gorakhargosh/watchdog.git
Cloning into 'watchdog'...
remote: Enumerating objects: 7723, done.
remote: Counting objects: 100% (551/551), done.
remote: Compressing objects: 100% (269/269), done.
error: object 0bc9713e4ae1664648f1b3fd1915638f28d66236: badTimezone: invalid author/committer line - bad time zone
fatal: fsck error in packed object
fatal: fetch-pack: invalid index-pack output

$ git --version
git version 2.35.1
```
Then if you change it back:
```
$ git config --global transfer.fsckobjects false

$ git clone https://github.com/gorakhargosh/watchdog.git
Cloning into 'watchdog'...
remote: Enumerating objects: 7723, done.
remote: Counting objects: 100% (551/551), done.
remote: Compressing objects: 100% (269/269), done.
remote: Total 7723 (delta 346), reused 422 (delta 274), pack-reused 7172
Receiving objects: 100% (7723/7723), 1.83 MiB | 5.07 MiB/s, done.
Resolving deltas: 100% (5112/5112), done.
[jschroeder@odysseus tmp]$ cd watchdog/
[jschroeder@odysseus watchdog] (master)$ git fsck
Checking object directories: 100% (256/256), done.
error in commit 0bc9713e4ae1664648f1b3fd1915638f28d66236: badTimezone: invalid author/committer line - bad time zone
error in commit a1cbf2de4aa9578db85b69e86dcd41e605fa56d5: badTimezone: invalid author/committer line - bad time zone
error in commit 33b8eb380fdb3d7712368037a8f49bc80f79b13f: badTimezone: invalid author/committer line - bad time zone
Checking objects: 100% (7723/7723), done.
```

Same author for all three of those invalid commits, so I guess the author had a very strange workstation configuration?

**Comment by kurtmckee** [comment]
I tried to script a fix for this. You'll lose git history but it will fix the corrupt commits.

```bash
# Clone the repository.
# The repository is corrupt, so fsck must be disabled.
git config --global transfer.fsckobjects false
git clone https://github.com/gorakhargosh/watchdog
cd watchdog/

# Checkout a new branch based on the commit
# just prior to the first corrupt commit.
git checkout -b fixed-commits 33b8eb380f^

# Cherry-pick each corrupt commit and fix its timezone.
for commit in 33b8eb380f a1cbf2de4a 0bc9713e4a; do
    ORIGINAL_MESSAGE="$(git show --quiet --pretty='%s' $commit)"
    ORIGINAL_DATE="$(git show --quiet --pretty='%ad' $commit)"
    NEW_DATE="$(echo $ORIGINAL_DATE | cut -d' ' -f1-5) +0000"
    echo "Cherry-picking $commit"
    echo "Original date: '$ORIGINAL_DATE'"
    echo "New date: '$NEW_DATE'"
    echo ""
    git cherry-pick $commit
    git commit --amend --date="$NEW_DATE" --message "$ORIGINAL_MESSAGE"
    echo ""
    echo ""
done

# Rebase master.
git checkout master
git rebase --onto fixed-commits 0bc9713e4a master --strategy ort -X ignore-space-change
```

You'll also need to delete tags and branches on the repo to ensure that the commits are unreachable, and then force-push the changes to GitHub. I recommend cloning this repository on GitHub and trying out these steps on a different repository first. You can run `git fsck --full` at various stages to confirm the results.

This may be an opportunity to rename the "master" branch at the same time.

**Comment by kurtmckee** [comment]
Makes sense; it's an extremely destructive, and likely not worth the lost history.

If there's nothing to be done for this ticket, it might be worthwhile to close it. :+1:
